File: day21/day21.cpp
#include <cassert>
#include <fstream>
#include <iostream>
#include <numeric>
#include <unordered_map>
#include <vector>

#include <fmt/format.h>

using namespace std;

static string rotate(const string& src)
{
	string dst(src);
	size_t l = src.find('/');
	size_t p = 0;
	for (size_t y = 0; y < l; y++)
	{
		for (size_t x = 0; x < l; x++)
		{
			dst[p++] = src[x*(l+1) + l-y-1];
		}
		p++;
	}
	return dst;
}

static string flip(const string &src)
{
	string dst(src);
	size_t l = src.find('/');
	size_t p = 0;
	for (size_t y = 0; y < l; y++)
	{
		for (size_t x = 0; x < l; x++)
		{
			dst[p++] = src[y*(l+1)+l-x-1];
		}
		p++;
	}
	return dst;
}

static unordered_map<string, string> load_rules(istream& in)
{
	unordered_map<string,string> rules;
	string src, dst;
	while (true)
	{
		in >> src;
		in >> dst;		// separator
		in >> dst;
		if (in.fail())
		{
			break;
		}
		for (int i = 0; i < 8; i++)
		{
			if (i == 4) src = flip(src);
			src = rotate(src);
			rules[src] = dst;
		}
	}
	return rules;
}

static string get_square(const vector<string>& map, size_t px, size_t py, size_t l)
{
	string dst;
	for (size_t y = py; y < py+l; y++)
	{
		for (size_t x = px; x < px+l; x++)
		{
			dst.push_back(map[y][x]);
		}
		dst.push_back('/');
	}
	dst.pop_back();
	return dst;
}

static void set_square(vector<string>& map, size_t px, size_t py, const string& dst)
{
	size_t l = dst.find('/');
	assert(l != string::npos);
	size_t p = 0;
	for (size_t y = py; y < py+l; y++)
	{
		for (size_t x = px; x < px+l; x++)
		{
			map[y][x] = dst[p];
			p++;
		}
		p++;
	}
}

static vector<string> step(const vector<string>& map,
			   const unordered_map<string,string>& rules)
{
	size_t srclen, dstlen;
	if (map.size() % 2)
	{
		srclen = 3;
		dstlen = 4;
	}
	else
	{
		srclen = 2;
		dstlen = 3;
	}

	size_t r = map.size() / srclen;
	vector<string> newmap(r*dstlen);
	for (auto& row: newmap)
	{
		row.resize(r*dstlen);
	}

	for (size_t y = 0; y < r; y++)
	{
		for (size_t x = 0; x < r; x++)
		{
			string src = get_square(map, x * srclen, y * srclen, srclen);
			auto dst = rules.find(src);
			if (dst != rules.end())
			{
				set_square(newmap, x * dstlen, y * dstlen, dst->second);
			}
			else
			{
				cerr << "rule not found for" << src << endl;
			}
		}
	}
	return newmap;
}

static size_t count_lit(const unordered_map<string,string>& rules, size_t rounds)
{
	vector<string> map = { ".#.", "..#", "###"};
	while (rounds-->0)
	{
		map = step(map, rules);
	}
	return accumulate(
		map.begin(),
		map.end(),
		0ULL,
		[](auto acc, const auto& row) {
			return acc + count(row.begin(), row.end(), '#');
		});
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		cerr << fmt::format("Usage: {} <filename>", argv[0]) << endl;
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		cerr << fmt::format("Cannot open {}", argv[1]) << endl;
		return 1;
	}
	auto rules = load_rules(input);
	input.close();

	cout << fmt::format("Part1: {}", count_lit(rules, 5)) << endl
	     << fmt::format("Part2: {}", count_lit(rules, 18)) << endl;
	return 0;
}

File: day21/day21.py
#!/usr/bin/env python3

import sys

def rotate(rule):
    table = [row for row in rule.split("/")]
    corner = len(table)
    ntable = []
    for y in range(corner):
        row = []
        for x in range(corner):
            row.append(table[x][len(table) - y - 1])
        ntable.append("".join(row))
    return "/".join(ntable)

def flip(rule):
    return "/".join("".join(reversed(row)) for row in rule.split("/"))

def parse(txt):
    rules = {}
    for row in txt.splitlines():
        src, dst = row.split(" => ")
        for i in range(8):
            if i == 4:
                src = flip(src)
            src = rotate(src)
            if src not in rules:
                rules[src] = dst
    return rules

def get_square(table, x, y, step):
    return "/".join(("".join(table[ty][tx] for tx in range(x, x+step)) for ty in range(y, y+step)))

def save_square(table, x, y, rule, rlen):
    pos = 0
    for ty in range(y, y+rlen):
        for tx in range(x, x+rlen):
            if rule[pos] == "/":
                pos += 1
            table[ty][tx] = rule[pos]
            pos += 1

def step(table, rules):
    corner = len(table)
    if corner % 2:
        srclen = 3
        dstlen = 4
    else:
        srclen = 2
        dstlen = 3

    r = corner // srclen
    ntable = [[' ' for i in range(r*dstlen)] for j in range(r*dstlen)]
    for ty in range(r):
        for tx in range(r):
            src = get_square(table, tx*srclen, ty*srclen, srclen)
            dst = rules[src]
            save_square(ntable, tx*dstlen, ty*dstlen, dst, dstlen)

    return ntable

def count_lit(rules, rounds):
    txt =""".#.
..#
###
"""
    table = [list(i) for i in txt.splitlines()]

    for i in range(rounds):
        table = step(table, rules)
    return sum("".join(row).count("#") for row in table)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    rules = parse(txt)
    print("Part1:", count_lit(rules, 5))
    print("Part2:", count_lit(rules, 18))
